lib/Sema/ConstraintSystem.cpp

```diff
@@ -4710,7 +4712,8 @@ void ConstraintSystem::maybeProduceFallbackDiagnostic(
   // diagnostics already emitted or waiting to be emitted. Because they are
   // a better indication of the problem.
   ASTContext &ctx = getASTContext();
-  if (ctx.hadError())
+  if (ctx.hadError() ||
+      (diagnosticTransaction && diagnosticTransaction->hasErrors()))
```

Comment:
It seems like `hadError()` is too restrictive here and we should add a new method on DiagnosticEngine i.e. `emittedOrHasPendingErrors()` which would examine TentativeDiagnostics from all currently open transactions as well as `hadError()`.

Reply:
I was thinking we should keep this narrow, so it doesn't get suppressed due to an unrelated diagnostic transaction.

Reply:
So I think my approach is better here than checking `TentativeDiagnostics`. I'm going to merge proactively, but am willing to change if I'm wrong.

Reply:
This approach checks only the innermost transaction, there could be others ones with errors.

Reply:
I think that's a feature. We don't know if we care about any other active transactions.